### PR TITLE
Added testkit support

### PR DESCRIPTION
**Pull request by LightningMods**

### CODE REVIEW

File: kernel/src/Mira.cpp [hunk 1]
```diff
@@ -295,8 +296,20 @@ bool Mira::Framework::Initialize()
 	// Set the running flag
 	m_InitParams.isRunning = true;
 
-  // Mira is now ready ! Now Killing SceShellUI for relaunching UI Process :D
-  struct proc* ui_proc = Mira::OrbisOS::Utilities::FindProcessByName("SceShellUI");
+struct proc* ui_proc = nullptr;
+// causes testkit IPC to kpanic the console on relaunch 
+//////////////////////////////////////////////////////////////
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
+if( OrbisOS::Utilities::isAssistMode() == IS_TESTKIT ||  OrbisOS::Utilities::isTestkit() == IS_TESTKIT){
```
Comment (kiwidoggie) [inline kernel/src/Mira.cpp]: Remove the #if statement, leave it isAssistMode() or isTestkit() to skip, else continue as normal.

File: kernel/src/OrbisOS/Utilities.cpp [hunk 2]
```diff
@@ -86,6 +86,52 @@ uint64_t Utilities::PtraceIO(int32_t p_ProcessId, int32_t p_Operation, void* p_D
         return (uint64_t)s_Desc.piod_len;
 }
 
+int Utilities::isAssistMode()
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.cpp]: make this bool

File: kernel/src/OrbisOS/Utilities.cpp [hunk 3]
```diff
@@ -86,6 +86,52 @@ uint64_t Utilities::PtraceIO(int32_t p_ProcessId, int32_t p_Operation, void* p_D
...
+{
+char out = -1;
+
+int descriptor = kopen_t(const_cast<char*>("/dev/dipsw"), 0, 0, curthread);
+if(descriptor >= 0)
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.cpp]: Invert the if statement to reduce indention

File: kernel/src/OrbisOS/Utilities.cpp [hunk 4]
```diff
@@ -86,6 +86,52 @@ uint64_t Utilities::PtraceIO(int32_t p_ProcessId, int32_t p_Operation, void* p_D
         return (uint64_t)s_Desc.piod_len;
 }
 
+int Utilities::isAssistMode()
+{
+char out = -1;
+
+int descriptor = kopen_t(const_cast<char*>("/dev/dipsw"), 0, 0, curthread);
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.cpp]: Use mira's main thread instead of current thread

File: kernel/src/OrbisOS/Utilities.cpp [hunk 5]
```diff
@@ -86,6 +86,52 @@ uint64_t Utilities::PtraceIO(int32_t p_ProcessId, int32_t p_Operation, void* p_D
...
+
+return -1;
+
+}
+
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
+int Utilities::isTestkit()
+{
+
+char target_id_addr = *(char *)kdlsym(target_id);
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.cpp]: I think this is a part of the large config block, will have to investigate further

File: kernel/src/OrbisOS/Utilities.cpp [hunk 6]
```diff
@@ -86,6 +86,52 @@ uint64_t Utilities::PtraceIO(int32_t p_ProcessId, int32_t p_Operation, void* p_D
...
+int testkit_mem_block[] = {0x82};
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.cpp]: This is unnneeded, just dereference the memory directly (you are already in kernel)

File: kernel/src/OrbisOS/Utilities.hpp [hunk 7]
```diff
@@ -5,6 +5,10 @@
 struct proc;
 struct thread;
 
+#define ASSIST_IOCTL 0x40048806
+#define IS_TESTKIT 1
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.hpp]: Remove this, it's not needed

File: kernel/src/OrbisOS/Utilities.hpp [hunk 8]
```diff
@@ -5,6 +5,10 @@
...
+#define IS_NOT_TESTKIT 0
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.hpp]: Same here

File: kernel/src/OrbisOS/Utilities.hpp [hunk 9]
```diff
@@ -41,6 +45,10 @@ namespace Mira
         public:
             static Utilities* GetInstance();
 
+            static int isAssistMode();
+ #if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.hpp]: Remove the 672 only options, this needs to be enabled for all versions

File: kernel/src/OrbisOS/Utilities.hpp [hunk 10]
```diff
@@ -41,6 +45,10 @@ namespace Mira
         public:
             static Utilities* GetInstance();
 
+            static int isAssistMode();
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.hpp]: bool return value

File: kernel/src/OrbisOS/Utilities.hpp [hunk 11]
```diff
@@ -41,6 +45,10 @@ namespace Mira
...
+ #if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
+            static int isTestkit();
```
Comment (kiwidoggie) [inline kernel/src/OrbisOS/Utilities.hpp]: bool return value

File: kernel/src/Plugins/FakePkg/FakePkgManager.cpp [hunk 12]
```diff
@@ -400,8 +400,21 @@ bool FakePkgManager::OnLoad()
         return false;
     }
 
+
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/Plugins/FakePkg/FakePkgManager.cpp]: Remove the version #if statement

File: kernel/src/Plugins/MorpheusEnabler/MorpheusEnabler.cpp [hunk 13]
```diff
@@ -122,7 +122,19 @@ bool MorpheusEnabler::OnLoad()
     m_processStartEvent = eventhandler_register(NULL, "process_exec_end", reinterpret_cast<void*>(MorpheusEnabler::ProcessStartEvent), NULL, EVENTHANDLER_PRI_LAST);
     m_resumeEvent = eventhandler_register(NULL, "system_resume_phase4", reinterpret_cast<void*>(MorpheusEnabler::ResumeEvent), NULL, EVENTHANDLER_PRI_LAST);
 
-	return DoPatch();
+
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/Plugins/MorpheusEnabler/MorpheusEnabler.cpp]: Remove the #if statement gating

File: kernel/src/Plugins/PluginManager.cpp [hunk 14]
```diff
@@ -113,14 +113,33 @@ bool PluginManager::OnLoad()
             break;
         }
 
-        // Initialize Substitute
-        m_Substitute = new Mira::Plugins::Substitute();
+
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/Plugins/PluginManager.cpp]: Remove the version gating

File: kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp [hunk 15]
```diff
@@ -36,8 +36,20 @@ void RemotePlayEnabler::ProcessStartEvent(void *arg, struct ::proc *p)
 
 	char* s_TitleId = (char*)((uint64_t)p + 0x390);
 
-	if (strncmp(s_TitleId, "NPXS20001", 9) == 0 && strcmp(p->p_comm, "SceShellUI") == 0)
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp]: Remove version gating

File: kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp [hunk 16]
```diff
@@ -47,7 +59,17 @@ void RemotePlayEnabler::ProcessStartEvent(void *arg, struct ::proc *p)
 
 void RemotePlayEnabler::ResumeEvent()
 {
-	ShellUIPatch();
+
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp]: Remove gating

File: kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp [hunk 17]
```diff
@@ -248,17 +270,34 @@ bool RemotePlayEnabler::OnLoad()
...
+
+auto s_Ret = RemotePlayPatch();
 	if (s_Ret == false) {
-		WriteLog(LL_Error, "could not patch SceShellUI");
+		WriteLog(LL_Error, "could not patch SceRemotePlay");
 		return false;
 	}
 
-	s_Ret = RemotePlayPatch();
+#if MIRA_PLATFORM==MIRA_PLATFORM_ORBIS_BSD_672
```
Comment (kiwidoggie) [inline kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp]: Remove gating

File: kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp [hunk 18]
```diff
@@ -248,17 +270,34 @@ bool RemotePlayEnabler::OnLoad()
 	m_processStartEvent = eventhandler_register(NULL, "process_exec_end", reinterpret_cast<void*>(RemotePlayEnabler::ProcessStartEvent), NULL, EVENTHANDLER_PRI_LAST);
 	m_resumeEvent = eventhandler_register(NULL, "system_resume_phase4", reinterpret_cast<void*>(RemotePlayEnabler::ResumeEvent), NULL, EVENTHANDLER_PRI_LAST);
 
-	auto s_Ret = ShellUIPatch();
+
+auto s_Ret = RemotePlayPatch();
```
Comment (kiwidoggie) [inline kernel/src/Plugins/RemotePlayEnabler/RemotePlayEnabler.cpp:267]: Fix indention

File: kernel/src/Utils/Kdlsym/Orbis672.hpp [hunk 19]
```diff
@@ -180,6 +180,7 @@
 #define kdlsym_addr_vsnprintf                              0x00123630
 #define kdlsym_addr_wakeup                                 0x0022A0A0
 #define kdlsym_addr_Xfast_syscall                          0x000001C0
+#define kdlsym_addr_target_id                              0x01BD800D
```
Comment (kiwidoggie) [inline kernel/src/Utils/Kdlsym/Orbis672.hpp:183]: Is this apart of the global config block?

File: kernel/src/Utils/Kernel.hpp [hunk 20]
```diff
@@ -32,19 +32,42 @@ typedef struct _ProcVmMapEntry
...
-	SceSysCore = 0x3800000000000007ULL,
-	Orbis_setip = 0x3800000000000008ULL,
-	GnmCompositor = 0x3800000000000009ULL,
-	SceShellUI = 0x380000000000000fULL, // NPXS20001
-	SceShellCore = 0x3800000000000010ULL,
-	NPXS20103 = 0x3800000000000011ULL,
-	NPXS21000 = 0x3800000000000012ULL,
-	// TODO: Fill in the rest
-	Decid = 0x3800000000010003,
+    sceSblACMgrIsAllowedToUsePupUpdate0 = 0x3800100000000001ULL,
```
Comment (kiwidoggie) [inline kernel/src/Utils/Kernel.hpp]: Sort these by value

**Review by kiwidoggie — CHANGES_REQUESTED** [review]
Ping me if have any questions, or leave a reply

File: kernel/src/OrbisOS/Utilities.cpp (hunk 5 above)
Comment (LightningMods) [reply kernel/src/OrbisOS/Utilities.cpp]: think so

File: kernel/src/OrbisOS/Utilities.cpp (hunk 6 above)
Comment (LightningMods) [reply kernel/src/OrbisOS/Utilities.cpp]: im not dereferencing it, im comparing it

**Pull request closed without merging**